### PR TITLE
Stralgo

### PR DESCRIPTION
**Pull request by AvitalFineRedis**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `$ tox` pass with this change (including linting)?
- [x] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [x] Is the new or changed code fully tested?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change

Add support to STRALGO command

### CODE REVIEW

**Comment by AvitalFineRedis** [comment]
@chayim

File: redis/client.py [hunk 1]
```diff
@@ -413,6 +413,34 @@ def parse_slowlog_get(response, **options):
...
+    Without modifiers the returned value is string.
+    When LEN is given the command returns the length of the result
+    (i.e integer).
+    When IDX is given the command returns a dictionary with the LCS
+    length and all the ranges in both the strings, start and end
+    offset for each string, where there are matches.
+    When WITHMATCHLEN is given, each array representing a match will
+    also have the length of the match at the beginning of the array.
+    """
+    if options['len']:
```
Comment (chayim) [inline redis/client.py:428]:
```suggestion
    if options.has_key'len':
```

File: redis/client.py [hunk 2]
```diff
@@ -413,6 +413,34 @@ def parse_slowlog_get(response, **options):
...
+        return int(response)
+    if options['idx']:
```
Comment (chayim) [inline redis/client.py:430]:
```suggestion
    if options.has_key('idx'):
```

File: redis/client.py [hunk 3]
```diff
@@ -413,6 +413,34 @@ def parse_slowlog_get(response, **options):
...
+        if options['withmatchlen']:
```
Comment (chayim) [inline redis/client.py:431]:
```suggestion
        if options.has_key('withmatchlen'):
```

File: redis/client.py [hunk 4]
```diff
@@ -2029,6 +2058,62 @@ def setrange(self, name, offset, value):
...
+
+        ``withmatchlen`` Returns the matches with the len of the match.
+        Can be provided only when ``idx`` set to True.
+        """
+        # check validity
+        supported_algo = ['LCS']
+        if algo not in supported_algo:
+            raise DataError("The supported algorithms are: %s"
+                            % (', '.join(supported_algo)))
+        if specific_argument:
```
Comment (chayim) [inline redis/client.py:2091]:
```suggestion
		if specific_argument is not None and \
		    specific_argument not in ['keys', 'strings']:
                raise DataError("specific_argument can be only"
                                " keys or strings")
```

File: redis/client.py [hunk 5]
```diff
@@ -2029,6 +2058,62 @@ def setrange(self, name, offset, value):
...
+            if specific_argument not in ['keys', 'strings']:
+                raise DataError("specific_argument can be only"
+                                " keys or strings")
+        else:
```
Comment (chayim) [inline redis/client.py:2095]: Remove this given the new default

File: redis/client.py [hunk 6]
```diff
@@ -2029,6 +2058,62 @@ def setrange(self, name, offset, value):
         """
         return self.execute_command('SETRANGE', name, offset, value)
 
+    def stralgo(self, algo, value1, value2, specific_argument=None, len=False,
```
Comment (chayim) [inline redis/client.py:2061]:
```suggestion
    def stralgo(self, algo, value1, value2, specific_argument=b'STRINGS', len=False,
```

File: redis/client.py [hunk 7]
```diff
@@ -2029,6 +2058,62 @@ def setrange(self, name, offset, value):
...
+                            % (', '.join(supported_algo)))
+        if specific_argument:
+            if specific_argument not in ['keys', 'strings']:
+                raise DataError("specific_argument can be only"
+                                " keys or strings")
+        else:
+            specific_argument = b'STRINGS'
+
+        pieces = [algo, specific_argument.upper(), value1, value2]
+        if len:
```
Comment (chayim) [inline redis/client.py:2099]:
```suggestion
        if len and idx:
```

File: redis/client.py [hunk 8]
```diff
@@ -2029,6 +2058,62 @@ def setrange(self, name, offset, value):
...
+            if idx:
+                raise DataError("len and idx cannot be provided together."
+                                " Just use idx.")
```
Comment (chayim) [inline redis/client.py:2102]: Remove this last line of the DataError. The first suffices.

File: redis/client.py (hunk 1 above)
Comment (AvitalFineRedis) [reply redis/client.py:428]: `has_key` will always return True.
Maybe `options.get('len', False)` will be better here @chayim ?

File: redis/client.py (hunk 2 above)
Comment (AvitalFineRedis) [reply redis/client.py:430]: Maybe options.get('idx', False) will be better here @chayim ?

File: redis/client.py (hunk 3 above)
Comment (AvitalFineRedis) [reply redis/client.py:431]: Maybe `options.get('withmatchlen', False)` will be better here @chayim ?

**Comment by AvitalFineRedis** [comment]
#1546

File: redis/client.py (hunk 1 above)
Comment (chayim) [reply redis/client.py:420]: Sure

File: redis/client.py (hunk 3 above)
Comment (chayim) [reply redis/client.py:423]: Sure

**Review by chayim — CHANGES_REQUESTED** [review]
@AvitalFineRedis Just the same changes. Then ready for merge.

**Comment by AvitalFineRedis** [comment]
@chayim  READY

**Pull request merged**